packages/ipfs-core/src/components/pin/remote/client.js

```diff
@@ -0,0 +1,283 @@
+'use strict'
+
+const CID = require('cids')
+const multiaddr = require('multiaddr')
+const HTTP = require('ipfs-utils/src/http')
+const log = require('debug')('ipfs:components:pin:remote:client')
+
+const withTimeoutOption = require('ipfs-core-utils/src/with-timeout-option')
+
+module.exports = ({ service, endpoint, key, swarm, peerId }) => {
+  const api = new HTTP({
+    base: endpoint,
+    headers: {
+      Authorization: `Bearer ${key}`,
+      'Content-Type': 'application/json'
+    }
+  })
+
+  async function info ({ stat: includeStats }) {
+    if (includeStats) {
+      return {
+        service,
+        endpoint: new URL(endpoint),
+        stat: await stat()
+      }
+    }
+    return { service, endpoint: new URL(endpoint) }
+  }
+
+  async function stat () {
+    try {
+      const promises = []
+      for (const pinStatus of ['queued', 'pinning', 'pinned', 'failed']) {
+        promises.push(countForStatus(pinStatus))
+      }
+      const [queued, pinning, pinned, failed] = await Promise.all(promises)
+      return {
+        status: 'valid',
+        pinCount: { queued, pinning, pinned, failed }
+      }
+    } catch (e) {
+      log(`error getting stats for service ${service}: `, e)
+      return {
+        status: 'invalid'
+      }
+    }
+  }
+
+  /**
+   * @param {string} status
+   * @returns {Promise<number>} - the number of remote pins with the given status
+   */
+  async function countForStatus (status) {
+    const searchParams = new URLSearchParams({ status, limit: '1' })
+    const response = await api.get('/pins', { searchParams })
+    const body = await response.json()
+    return body.count
+  }
+
+  async function originAddresses () {
+    const addrs = await swarm.localAddrs()
+    return addrs.map(ma => {
+      const str = ma.toString()
+
+      // some relay-style transports add our peer id to the ma for us
+      // so don't double-add
+      if (str.endsWith(`/p2p/${peerId}`)) {
+        return str
+      }
+
+      return `${str}/p2p/${peerId}`
+    })
+  }
+
+  async function connectToDelegates (delegates) {
+    const addrs = delegates.map(multiaddr)
+    const promises = []
+    for (const addr of addrs) {
+      promises.push(swarm.connect(addr).catch(e => {
+        log('error connecting to pinning service delegate: ', e)
+      }))
+    }
+    await Promise.all(promises)
+  }
+
+  async function awaitPinCompletion (pinResponse) {
+    const pollIntervalMs = 100
+
+    let { status, requestid } = pinResponse
+    while (status !== 'pinned') {
+      if (status === 'failed') {
+        throw new Error('pin failed: ' + JSON.stringify(pinResponse.info))
+      }
+
+      await delay(pollIntervalMs)
```

Comment:
I wonder if we should make `pollInterval` part of the optional `add` options so it could be configured if necessary to longer / shorter interval.